File: Batch4ProblemSolving/powx-n.java
//https://leetcode.com/problems/powx-n/solution/

class Solution {
    public double myPow(double x, int n) {
        if(n < 0) return myPow(1/x, -n);
        else if(n == 0) return 1.0;
        else if(n == 1) return x;
        else if(n % 2 == 0) return myPow(x * x, n / 2);
        else  return myPow(x * x, n / 2) * x;
    }
}

File: Batch4ProblemSolving/powx-n.kt
//https://leetcode.com/problems/powx-n/solution/
class Solution {
    fun myPow(x: Double, n: Int): Double {
        var a: Double = x
        var y: Long = n.toLong()
        
        if(y < 0){
            a = 1 / a
            y = -y
        }
        
        var ans: Double = 1.0
        var prod: Double = a
        
        while(y > 0) {
            if(y % 2 == 1L) ans = ans * prod
            //println(ans)
            prod = prod * prod
            y = y / 2
        }
        
        return ans
    }
}
